[24283] Wiki macros now accept an optional block of text for further processing

silkentrance opened the pull request:
Wiki macros now accept an optional block of text for further processing

- This is basically a port of the same functionality from redmine to OP.
- Some refactoring was required to make this work. The interface of macro functions has changed. They must now expect a ``**options`` parameter. By default, a ``block:`` which represents the block of text from the parsed macro, will be passed as a named parameter to the macro.
- Additional refactoring was required to solve errors reported by codeclimate/rubocop.

Still working on test cases, so consider this a WIP.

https://community.openproject.com/work_packages/24283

oliverguenther commented:
Intention-wise, I agree with this change. 👍 

I'm always be a little scared touching the text formatting code since it is so unnecessary complex and brittle, so props for trying 😉 
Drop me a mail if you have questions regarding tests or other parts of th e application.

silkentrance commented:
@oliverguenther thanks. yes, the code is quite complex and quite fragile, order of processing and the order of markup itself is really important.

silkentrance commented:
See https://github.com/opf/openproject/pull/5074#issuecomment-262838282 for more information. WIP DO NOT MERGE.

silkentrance commented:
Woah during the merge all local changes have been lost, git stash pop on the loose.... Need to redo this one more time.

silkentrance committed:
Wiki macros now accept an optional block of text for further processing

https://community.openproject.com/work_packages/24283

silkentrance commented:
@oliverguenther fixed, merged past changes back into newly created branch from release/6.1.
got rid of ``parse_non_pre_blocks``. It also incorporates the change where macros now need to be notated as

```
{{{macro...}}}
```

instead of using two curly braces. The rationale here is that we need to get rid of the ``escape_non_macros`` stage so that overall processing speed increases. In addition, this will eradicate all false positives marked as failures when trying to execute something such as ``{{project.attr}}`` and that all the special casing in the frontend regarding DOUBLE_LEFT_CURLY_BRACE can be eliminated as well.

*Notes*

- this also includes two additional built-in (dummy) macros, namely pre and code from which one can build upon.

- currently works also recursively for single line macros

- recursive operation on multi line macros requires a separate parser

- macro_list is broken and does not render correctly

- built in macros require more work, @project is nil

silkentrance commented:
@oliverguenther fixed, merged past changes back into newly created branch from release/6.1.

oliverguenther commented:
@silkentrance As much as I wholeheartedly agree with avoiding using angular expressions for macros, this will cause one hell of a migration problem since lots of production textile uses `{{macro}}`. 

We can also not avoid `DOUBLE_LEFT_CURLY_BRACE` since we do not want arbitrary angular expressions to be executed. We do have some ruby templates where expressions are used, so we can't `ng-non-bindable`the whole body either. Angular will also shut down its sandbox in 1.6.

silkentrance commented:
@oliverguenther Ah, ok, I see. So it is double curly braces then. I will revert this change in the next push.
Here, however, is a quick preview on the new parser that supports nesting of macro calls and, provided that the macro being passed the text that contains additional macro calls, it is free to evaluate them using the standard formatter or work its own magic from there.

As for the pre macro, this is now straigt forward and requires no special casing whatsoever. The pre macro will not call upon the formatter but instead will render a html safe version of the text.

The same goes for code, however, here one can have additional parameters that will then be used to parameterize the code element.


``parse_non_pre_blocks`` has been replaced by ``pre_process_pre_blocks`` which in turn will replace all occurrences of ``<pre>|<code>`` by a macro call. This is far from being optimal as any parameters (attributes) are lost including also the inability to have ``<pre>block</pre>`` single line pre blocks, as ``pre_process_pre_blocks`` is quite simple. Also, the step of closing any unmatched tags is lost and needs to be implemented.

I know that a lot of existing tests are failing due to these changes and this will be addressed once we can come to a conclusion on how to handle these pre/code blocks, if you will.

silkentrance commented:
The latest push now features

- pre/code blocks will be migrated from original html based notation to a macro based notation
  html attributes are passed as named parameters to macro, macros should make use of Nokogiri instead
- macros will not be evaluated inside pre/code blocks, which is due to the way that the macros are implemented
- macro code can be nested arbitrarily, recursive macro processing works
- TODO:code blocks inside pre will not be evaluated, e.g.

```
<pre>
<code>
</code>
</pre>
```

will be evaluated to

```
<pre>
{{code()
}}
</pre>
```

oliverguenther commented:
I'm not yet convinced of the benefits that migrating pre|code blocks to a macro will have. At the moment, this change increases the complexity of text_formatting rather than reduce it.

What other incentive do you have for multiline and recursive macros except for these two blocks?

silkentrance commented:
The main reason I did this was to be able to extract the macros and then inject them again in a separate step. But I think that the two steps can be merged and made part of the old logic. I will revise this in the next push.

silkentrance commented:
@oliverguenther This required some refactoring in order to get it right. While I was at it, I also refactored out the argument parsing to a separate method. In addition, headings are now part of a post-processing step, hopefully both simplifying and optimizing the performance of the existing code.

Parsing the ``{{toc}}`` is now optional, and by default, false is assumed. Therefore the change to the wiki's  ``_content.html.erb`` which must now call ``format_text`` with the required ``headings`` option. And, in case that the ``headings`` option is false, all occurrences of ``{{toc}}`` will be removed, which is helpful on ``{{include(...)}}``.

Use of ``SecureRandom.hex`` was introduced to make it near impossible for the user to guess injected macro names.

Added boolean option ``macros`` which controls, in addition to ``edit``, whether macros should be executed.

This also includes further refactorings I found necessary and which might help reduce the complexity (verbosity/unused parameters) of existing interfaces.


MACROS_RE and MACROS_EVAL_RE should be made one, but I am not able to figure out a way on how to do that. MACROS_RE matches arguments to embedded macros, where it should not, e.g.

```
{{macro
{{include(child)}}
}}
```

will cause MACRO_RE to match args of the include macro and map these to the outer ``macro`` macro, which of course is wrong. Whereas the MACROS_EVAL_RE does not do recursive matches but instead will return the correct args for the outer macro, which are the empty string.

I have tried several things, e.g. matching for ``[(&&^\n\r]`` but this will cause the whole shebang to fail and only inner macros will be detected.

Perhaps you can find a way?

silkentrance commented:
@oliverguenther the second commit contains further refactorings I found necessary and which reduce the complexity (brevity) of existing interfaces.

oliverguenther commented:
Wouldn't it be possible to restrict arguments only on the same line the macro is opened? That way we could reduce the regex to match each macro only on a single line. A bigger issue I believe would be correctly matching each closing bracket correctly. At least I don't think this happens in the current regex.

Maybe we should rejoin to think again about a way moving away from the brackets, but provide a backwards compatible parser for the previous syntax? If you start allowing multiple nested macros, I'd believe that input something of the form

```
<include page=foo>
<foo-macro></foo-macro>
</include>
```

would be easier to match with less manual labor on our side. Even with nested macros of the same kind, there are existing parsers that do this for us.

silkentrance commented:
Superceded by #5133
